### PR TITLE
Fix batch status flickering

### PR DESCRIPTION
**Pull request by amlcurran**
This PR adds fixes an issue where, when a download in a batch completes, the batch's status would briefly go to "queued", even though from the batch's point of view it is still running.

Also, it looks like the notification doesn't flicker either!

FYI the flicker happens at around 10MB

| Before | After |
| --- | --- |
| broken-wobble | fix-wobbly-state |

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import android.content.ContentResolver;
 import android.content.ContentValues;
 import android.database.Cursor;
-import android.util.SparseIntArray;
+import android.support.v4.util.SparseArrayCompat;
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: Using the compat version allows us to test the functionality, although does mean we get an autoboxing penalty.

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/159/
Test FAILed.

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java [hunk 2]
```diff
@@ -86,8 +98,16 @@ int getBatchStatus(long batchId) {
             }
         }
 
+        boolean hasCompleteItems = statusCountMap.hasCountFor(DownloadStatus.SUCCESS);
+        boolean hasPendingItems = statusCountMap.hasCountFor(DownloadStatus.PENDING);
+        boolean hasSubmittedItems = statusCountMap.hasCountFor(DownloadStatus.SUBMITTED);
+        boolean hasOtherItems = statusCountMap.hasNoItemsWithStatuses(STATUSES_EXCEPT_PENDING_SUCCESS_SUBMITTED);
+        if (hasCompleteItems && (hasPendingItems || hasSubmittedItems) && !hasOtherItems) {
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: this is kinda grim, sorry.

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/160/
Test PASSed.

File: demo-extended/src/main/java/com/novoda/downloadmanager/demo/extended/batches/BatchesActivity.java [hunk 3]
```diff
@@ -20,7 +19,6 @@
 public class BatchesActivity extends AppCompatActivity implements QueryForBatchesAsyncTask.Callback {
 
     private final Handler handler = new Handler(Looper.getMainLooper());
-    private final QueryTimestamp lastQueryTimestamp = new QueryTimestamp();
```
Comment (xrigau) [inline demo-extended/src/main/java/com/novoda/downloadmanager/demo/extended/batches/BatchesActivity.java]: :+1: forgot to get rid of this

File: library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    public void givenABatchWithSomePendingItemsAndSomeCompleteItemsAndSomeSubmittedItemsThenTheBatchStatusIsRunning() {
+        BatchRepository repository = givenABatchWithStatuses(DownloadStatus.PENDING, DownloadStatus.SUCCESS, DownloadStatus.SUBMITTED);
+
+        int batchStatus = repository.getBatchStatus(1);
+
+        assertThat(batchStatus).isEqualTo(DownloadStatus.RUNNING);
+    }
+
+    @Test
+    public void givenABatchWithAMixOfItemsThenTheBatchStatusIsNotRunning() {
```
Comment (xrigau) [inline library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java]: previous test name is really specific whereas this one isn't :-1: I had to scroll to see that 'a mix of items' means there's one failed item. why not using a more specific name here?

Comment (amlcurran) [inline library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java]: I couldn't think of one that wasn't really long :( but I'll flip the test name around

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java [hunk 5]
```diff
@@ -183,4 +203,35 @@ private void deleteBatchesForIds(List<Long> batchIdsToDelete, Collection<FileDow
...
+                if (hasItemsForStatus) {
+                    hasOtherItems = true;
+                    break;
+                }
+            }
+            return hasOtherItems;
+        }
+
+        public boolean hasCountFor(int status) {
+            return statusCounts.get(status, 0) > 0;
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: maybe you can avoid some unboxing by using the `compareTo` method in the `Integer` class
http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#compareTo(java.lang.Integer)

http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/lang/Integer.java#Integer.compareTo%28java.lang.Integer%29

Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: but don't think it's a bottleneck

File: library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java (hunk 4 above)
Comment (amlcurran) [inline library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java]: https://github.com/novoda/download-manager/commit/ef5d3721afb030a80f1b46acc8d79e9b2c81cd06

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/161/
Test PASSed.

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java (hunk 5 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: We have the autoboxing as soon as we add something to the array. In Android OS there is a SparseIntArray which is backed by an `int[]` but there is no equivalent in the support library

Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: :(

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java (hunk 2 above)
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: I don't understand why we need to check if there's complete items?

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java [hunk 6]
```diff
@@ -183,4 +203,35 @@ private void deleteBatchesForIds(List<Long> batchIdsToDelete, Collection<FileDow
...
+    private static class StatusCountMap {
+
+        private final SparseArrayCompat<Integer> statusCounts = new SparseArrayCompat<>(PRIORITISED_STATUSES_SIZE);
+
+        public boolean hasNoItemsWithStatuses(List<Integer> excludedStatuses) {
+            boolean hasOtherItems = false;
+            for (int status : excludedStatuses) {
+                boolean hasItemsForStatus = hasCountFor(status);
+                if (hasItemsForStatus) {
+                    hasOtherItems = true;
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: could just `return true` here and on L220 just `return false`

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java (hunk 5 above)
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: yeah I saw that :disappointed:

File: library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java (hunk 4 above)
Comment (xrigau) [inline library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusTests.java]: :+1:

**Comment by xrigau** [comment]
:+1: in general LGMT, just waiting on the Q about the completed status

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java (hunk 2 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: If you have 1 complete and 2 pending items, the batch is technically running (you've just finished one download before starting the next), but the batch's status will actually be pending. 

But if it has no complete items you're at the start of the download and it is actually pending

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java (hunk 6 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java (hunk 2 above)
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java]: It makes a bit more sense but I don't fully understand 100%, anyway I trust you and the tests so :+1: from me

**Comment by xrigau** [comment]
will :ship:  if no other comments

**Comment by ouchadam** [comment]
:+1: seems legit but I do worry if it will break something with some more extensive testing. let's find out!

**Comment by amlcurran** [comment]
it's not like we've worried about that with our other PRs :tongue: 

But I have tested it in other places and it works OK :+1:

**Comment by xrigau** [comment]
:roller_coaster:

**Comment by Electryc** [comment]
> Also, it looks like the notification doesn't flicker either!

Winning.

**Comment by blundell** [comment]
> it's not like we've worried about that with our other PRs

we're always worried

**Comment by xrigau** [comment]
> > it's not like we've worried about that with our other PRs

> we're always worried

(⊙﹏⊙)

**Pull request merged**